### PR TITLE
Optimizer

### PR DESCRIPTION
**Pull request by paquiteau**

### CODE REVIEW

**Comment by paquiteau** [comment]
Don't Worry, the test are coming with the PR for reconstructors.

**Comment by chaithyagr** [comment]
Added @Daval-G for review.

File: mri/optimizers/base.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+def run_online_algorithm(opt, kspace_generator, estimate_call_period, verbose=0):
+    """Run online optimisation algorithm.
+
+    At each step the obs_data is updated viw the kspace_generator.
+
+    Parameters
+    ----------
+    opt: optimisation algorithm instance
+    kspace_generator: kspace data
+    estimate_call_period: int, default None
```
[the remark this answers is not in the document]
Comment (Daval-G) [reply mri/optimizers/base.py:65]: Also it would be nice to know the `kspace_generator` and `opt` types.

File: mri/optimizers/utils/cost.py [hunk 2]
```diff
@@ -143,7 +152,12 @@ def _calc_cost(self, x_new, *args, **kwargs):
             the cost function defined by the operators (gradient + prox_op).
         """
         if self.optimizer_type == 'forward_backward':
-            cost = self.gradient_op.cost(x_new) + self.prox_op.cost(x_new)
+            if not hasattr(self.grad_op, 'linear_op') and self.linear_op is not None:
+                y_new = self.linear_op.op(x_new)
+            else:
+                # synthesis case, this already in the linear_op (sparse) domain.
```
Comment (Daval-G) [inline mri/optimizers/utils/cost.py:158]: Missing words

**Review by Daval-G — CHANGES_REQUESTED** [review]
It looks good to me, aside from minor typos. However it would be nice to add some minimal tests.

**Comment by paquiteau** [comment]
To add test on the online algorithms similar to the existing one for the offline algorithms I will need to add the [OnlineReconstructor](https://github.com/paquiteau/pysap-mri-online/blob/master/online/reconstructors/reconstructor.py), which need  the generator PR to be merge, and Ideally my PR on Modopt for gradient descent algorithms, to have a complete picture.

**Comment by chaithyagr** [comment]
> To add test on the online algorithms similar to the existing one for the offline algorithms I will need to add the [OnlineReconstructor](https://github.com/paquiteau/pysap-mri-online/blob/master/online/reconstructors/reconstructor.py), which need the generator PR to be merge, and Ideally my PR on Modopt for gradient descent algorithms, to have a complete picture.

In my opinion, we can wait for those PRs to go through. In the meantime, @paquiteau do keep the tests ready. Once other PRs are done, we can test finally and submit. Sounds good?

**Comment by chaithyagr** [comment]
other option is to write it as TODO, open a new issue and make sure it is done later.

**Review by Daval-G — APPROVED** [review]
Seems good to merge, besides some quick comment to address

**Pull request merged**